### PR TITLE
[Executor] Provide exec_aggregation_async for script executor

### PR DESCRIPTION
**Pull request by lumoslnt**
# Description

Provide exec_aggregation_async for script executor.

# All Promptflow Contribution checklist:
- [x] **The pull request does not introduce [breaking changes].**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [ ] **Create an issue and link to the pull request to get dedicated review from promptflow team. Learn more: [suggested workflow](../CONTRIBUTING.md#suggested-workflow).**

## General Guidelines and Best Practices
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: src/promptflow-core/promptflow/executor/_script_executor.py [hunk 1]
```diff
@@ -187,6 +187,34 @@ def _exec_aggregation(
...
+            return await self._exec_aggregation_async(aggregation_inputs)
+
+    async def _exec_aggregation_async(self, inputs):
+        output = None
+        try:
+            if inspect.iscoroutinefunction(self._aggr_func):
+                output = await self._aggr_func(**{self._aggr_input_name: inputs})
+            else:
+                partial_func = partial(self._aggr_func, **{self._aggr_input_name: inputs})
+                output = await asyncio.get_event_loop().run_in_executor(None, partial_func)
```
Comment (thy09) [inline src/promptflow-core/promptflow/executor/_script_executor.py:210]: I guess we may prepare `self._aggr_func` and `self._async_aggr_func` when initialization so we could make the logic here more clear?

Comment (thy09) [inline src/promptflow-core/promptflow/executor/_script_executor.py:210]: We may need to copy the parent context to the child thread. Using promptflow.tracing.ThreadPoolExecutorWithContext

Comment (lumoslnt) [reply src/promptflow-core/promptflow/executor/_script_executor.py]: Refined. Will do the same thing for exec line.

Comment (lumoslnt) [reply src/promptflow-core/promptflow/executor/_script_executor.py]: Refined.

File: src/promptflow-core/promptflow/executor/_script_executor.py [hunk 2]
```diff
@@ -348,7 +370,21 @@ def _initialize_aggr_function(self, flow_obj: object):
...
+                def run_async_function_sync(*args, **kwargs):
+                    return async_run_allowing_running_loop(aggr_func, *args, **kwargs)
+
+                self._aggr_func = run_async_function_sync
+                self._aggr_func_async = aggr_func
+            else:
+
+                async def run_sync_function_async(*args, **kwargs):
+                    with ThreadPoolExecutorWithContext() as executor:
+                        return executor.submit(aggr_func, *args, **kwargs).result()
```
Comment (thy09) [inline src/promptflow-core/promptflow/executor/_script_executor.py:384]: please use `await run_in_loop` to avoid blocking the world.

Comment (lumoslnt) [reply src/promptflow-core/promptflow/executor/_script_executor.py]: Fixed.

File: src/promptflow-core/promptflow/executor/_script_executor.py [hunk 3]
```diff
@@ -176,10 +177,31 @@ def _exec_aggregation(
     ) -> AggregationResult:
         output = None
         try:
-            if inspect.iscoroutinefunction(self._aggr_func):
-                output = async_run_allowing_running_loop(self._aggr_func, **{self._aggr_input_name: inputs})
-            else:
-                output = self._aggr_func(**{self._aggr_input_name: inputs})
+            output = self._aggr_func(**{self._aggr_input_name: inputs})
+            metrics = output if isinstance(output, dict) else {"metrics": output}
```
Comment (thy09) [inline src/promptflow-core/promptflow/executor/_script_executor.py:181]: No need to have "metrics?"

Comment (thy09) [reply src/promptflow-core/promptflow/executor/_script_executor.py:181]: Let's use another PR for it.

**Review by thy09 — APPROVED** [review]

**Review by brynn-code — APPROVED** [review]

**Pull request merged**